Writerside/topics/results-and-discussion/testing.md

```diff
@@ -0,0 +1,42 @@
+hand, assess multiple components interacting together (e.g., a file reader working with a records parser and test data).
+In our case, integration tests may also include running external dependencies, such as in the client, where a service
+instance must be started using Docker Compose. We acknowledge that the boundary between integration and system tests can
+be blurred, but we have chosen to categorize these as integration tests rather than introducing a third test category.
+It is important that unit tests and integration tests can be executed in distinct phases.
+
+We did not prioritize achieving the highest possible line coverage. Instead, we used this metric as a guiding tool
+during development. The focus was on testing the most relevant and complex features. During the review process,
+reviewers were required to check the code against the [Code Style Guide](code-style-guide.md), which also included
+ensuring that new or modified components were properly tested.
+
+## Setup
+
+- **Java, Maven**: We use JUnit 5, the Maven Surefire Plugin for unit tests (executed in the Maven `test` phase, with
+  a `<TestName>Test` suffix), and the Maven Failsafe Plugin (executed in the Maven `verify` phase, with a `<TestName>IT`
+  suffix) for integration tests. This allows unit and integration tests to be run in separate phases.
+
+- **Python, Poetry**: Pytest is used as the test runner, with Pytest markers to distinguish between unit and integration
+  test categories.
+
+- **C++, CMake and vcpkg**: GoogleTest is used for testing (TODO).
+
+## Code Metrics
+
+The following table provides an overview of some key code metrics for the main repositories involved in this thesis. The
+lines of code were counted using the command-line tool `cloc` and include both production and test code, excluding
+inline comments and documentation (such as Javadoc or docstrings). Test coverage is measured in terms of line coverage.
+
+| Repository Name            | Language | Short Description                                                                   | Commits | Lines of Code *(Test)* | Test Cases *(Integration)* | Test Coverage |
+|----------------------------|----------|-------------------------------------------------------------------------------------|---------|------------------------|----------------------------|---------------|
+| **public-transit-service** | Java     | Public transit schedule and routing service based on GTFS data and RAPTOR algorithm | 751     | 12'116 *(5'662)*       | 569 *(51)*                 | 86%           |
+| **public-transit-client**  | Python   | Client to access the public transit service API endpoints.                          | 72      | 822 *(419)*            | 23 *(15)*                  | 80%           |
+| **public-transit-viewer**  | Python   | Viewer to interact with the public transit service.                                 | 125     | 1'007 *(21)*           | 1 *(1)*                    | 0%            |
```

Comment:
how can we have 0% coverage?

Reply:
I dont know 😄 It seems like our test is mostly testing empty __init__.py files:

I guess this is due to the subprocess we are starting, and therefore pytest cannot follow the lines anymore (since in another process):

```python
    process = subprocess.Popen(
        ["python3", "-m", "public_transit_viewer.entrypoint"],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
```

Coverage results:

```
❯ poetry run pytest --cov=public_transit_viewer
=================================================================== test session starts ====================================================================
platform linux -- Python 3.12.5, pytest-8.3.3, pluggy-1.5.0
rootdir: /mnt/c/devsbb/code/naviqore/public-transit-viewer
configfile: pyproject.toml
plugins: cov-5.0.0
collected 1 item

tests/test_entrypoint.py ./mnt/c/devsbb/code/naviqore/public-transit-viewer/.venv/lib/python3.12/site-packages/coverage/control.py:894: CoverageWarning: No data was collected. (no-data-collected)
  self._warn("No data was collected.", slug="no-data-collected")
                                                                                                                           [100%]

---------- coverage: platform linux, python 3.12.5-final-0 -----------
Name                                                  Stmts   Miss  Cover
-------------------------------------------------------------------------
public_transit_viewer/__init__.py                         3      3     0%
public_transit_viewer/assets/__init__.py                  0      0   100%
public_transit_viewer/components/__init__.py              0      0   100%
public_transit_viewer/components/form_components.py      69     69     0%
public_transit_viewer/components/logo.py                  7      7     0%
public_transit_viewer/components/map.py                  78     78     0%
public_transit_viewer/connections.py                     30     30     0%
public_transit_viewer/entrypoint.py                      10     10     0%
public_transit_viewer/pages/__init__.py                   0      0   100%
public_transit_viewer/pages/isolines.py                  80     80     0%
public_transit_viewer/utils/__init__.py                   0      0   100%
public_transit_viewer/utils/client.py                   115    115     0%
public_transit_viewer/utils/color.py                     14     14     0%
public_transit_viewer/utils/connection.py                99     99     0%
-------------------------------------------------------------------------
TOTAL                                                   505    505     0%


==================================================================== 1 passed in 2.71s =====================================================================
```

Reply:
Added note

Reply:
https://github.com/naviqore/documentation/pull/25/commits/f395a3d852abf63e0c1a458e432477ae70524c28